src/MOIWrapper.jl

```diff
@@ -0,0 +1,345 @@
+    for i in 1:length(ci)
+        f = MOI.get(userOptimizer, MOI.ConstraintFunction(), ci[i])
+        push!(zeroOneIndices, mapping.varmap[f.variable].value)
+    end
+end
+
+function updateIntegerIndices(userOptimizer::MOI.ModelLike, mapping::MOIU.IndexMap,
+    ci::Vector{MOI.ConstraintIndex{F,S}}, integerIndices::Vector{Int}) where {F, S}
+    for i in 1:length(ci)
+        f = MOI.get(userOptimizer, MOI.ConstraintFunction(), ci[i])
+        push!(integerIndices, mapping.varmap[f.variable].value)
+    end
+end
+
+    """
+    Receive a cbcOptimizer which contains the pointer to the cbc C object and instanciate the object cbcModelFormat::CbcModelFormat based on userOptimizer::AbstractModel (also provided by the user).
+    Function loadProblem of CbcCInterface requires all information stored in cbcModelFormat.
+    """
+function MOI.copy!(cbcOptimizer::CbcOptimizer,
+    userOptimizer::MOI.ModelLike; copynames=false)
+
+    mapping = MOIU.IndexMap()
+
+    nbCols = MOI.get(userOptimizer, MOI.NumberOfVariables())
+    varIndex = MOI.get(userOptimizer, MOI.ListOfVariableIndices())
+    for i in 1:nbCols
+        mapping.varmap[varIndex[i]] = MOI.VariableIndex(i)
+    end
+
+    zeroOneIndices = Int[]
+    integerIndices = Int[]
+    listOfConstraints = MOI.get(userOptimizer, MOI.ListOfConstraints())
+    nbRows = 0
+    for (F,S) in listOfConstraints
+        if !(MOI.supportsconstraint(cbcOptimizer, F, S))
+            return MOI.CopyResult(MOI.CopyUnsupportedConstraint,
+            "Cbc MOI Interface does not support constraints of type " * (F,S) * ".", nothing)
+        end
+
+        ci = MOI.get(userOptimizer, MOI.ListOfConstraintIndices{F,S}())
+
+        if S == MOI.ZeroOne
+            updateZeroOneIndices(userOptimizer, mapping, ci, zeroOneIndices)
+        elseif S == MOI.Integer
+            updateIntegerIndices(userOptimizer, mapping, ci, integerIndices)
+        end
+
+        ## Update conmap for (F,S) for F != MOI.SingleVariable
+        ## Single variables are treated by bounds in Cbc, so no
+        ## need to add a row
+        if F != MOI.SingleVariable
+            for i in 1:length(ci)
+                mapping.conmap[ci[i]] = MOI.ConstraintIndex{F,S}(nbRows + i)
+            end
+            nbRows += MOI.get(userOptimizer, MOI.NumberOfConstraints{F,S}())
+        end
+    end
+
+    cbcModelFormat = CbcModelFormat(nbRows, nbCols)
+
+    copyconstraints!(cbcModelFormat, userOptimizer, mapping)
+    updateBoundsForBinaryVars!(cbcModelFormat.col_lb, cbcModelFormat.col_ub, zeroOneIndices)
+
+
+    ## Copy objective function
+    objF = MOI.get(userOptimizer, MOI.ObjectiveFunction{MOI.ScalarAffineFunction{Float64}}())
+    loadObj(cbcModelFormat, mapping, objF)
+    sense = MOI.get(userOptimizer, MOI.ObjectiveSense())
+    MOI.set!(cbcOptimizer, MOI.ObjectiveSense(), sense)
+
+    ## Load the problem to Cbc
+    CbcCI.loadProblem(cbcOptimizer.inner, cbcModelFormat.constraint_matrix,
```

Comment:
Do we need to keep the `constraint_matrix` in memory after it's loaded into Cbc?

Reply:
This object is local to the function `copy!` where it is defined, isn't it?

Reply:
You're right, I missed that.